File: Ahorn/entities/TitanStatue.jl
module TitanStatue

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/TitanStatue" TitanStatue(x::Integer, y::Integer, canAwaken::Bool=false, health::Integer=5, isAnimated::Bool=false)

const placements = Ahorn.PlacementDict(
    "inactive" => Ahorn.EntityPlacement(TitanStatue),
    "can_awaken" => Ahorn.EntityPlacement(TitanStatue),
    "animated" => Ahorn.EntityPlacement(TitanStatue)
)

function Ahorn.selection(entity::TitanStatue)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TitanStatue, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/titanStatue.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local titanStatue = {}

titanStatue.name = "MaggyHelper/TitanStatue"
titanStatue.depth = -100
titanStatue.fieldInformation = {
    health = {
        minimumValue = 1
    }
}
titanStatue.placements = {
    {
        name = "inactive",
        data = {
            health = 5,
            canAwaken = false,
            isAnimated = false
        }
    },
    {
        name = "can_awaken",
        data = {
            health = 5,
            canAwaken = true,
            isAnimated = false
        }
    },
    {
        name = "animated",
        data = {
            health = 5,
            canAwaken = true,
            isAnimated = true
        }
    }
}

function titanStatue.sprite(room, entity)
    local state = entity.isAnimated and "active" or "inactive"
    local sprite = drawableSprite.fromTexture("objects/titan_statue/" .. state, entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    return sprite
end

function titanStatue.selection(room, entity)
    return utils.rectangle(entity.x - 24, entity.y - 80, 48, 80)
end

return titanStatue
